engine/src/main/java/org/pentaho/di/trans/steps/mergejoin/MergeJoin.java

```diff
@@ -0,0 +1,466 @@
+          putRow( data.outputRowMeta, outputRowData );
+        }
+        data.one = data.one_next;
+        data.two = data.two_next;
+        break;
+      case 1:
+        // if (log.isDebug()) logDebug("First stream has missing key");
+        /*
+         * First stream is greater than the second stream. This means: a) This key is missing in the first stream b)
+         * Second stream may have finished So, if full/right outer join is set and 2nd stream is not null, we push a
+         * record to output with only the values for the second row populated. Next, if 2nd stream is not finished, we
+         * get a row from it; otherwise signal that we are done
+         */
+        if ( data.one_optional == true ) {
+          if ( data.two != null ) {
+            Object[] outputRowData = RowDataUtil.createResizedCopy( data.one_dummy, data.outputRowMeta.size() );
+            outputRowData = RowDataUtil.addRowData( outputRowData, data.oneMeta.size(), data.two );
+            putRow( data.outputRowMeta, outputRowData );
+            data.two = getRowFrom( data.twoRowSet );
+          } else if ( data.two_optional == false ) {
+            /*
+             * If we are doing right outer join then we are done since there are no more rows in the second set
+             */
+            // Before we stop processing, we have to make sure that all rows from both input streams are depleted!
+            // If we don't do this, the transformation can stall.
+            //
+            while ( data.one != null && !isStopped() ) {
+              data.one = getRowFrom( data.oneRowSet );
+            }
+            while ( data.two != null && !isStopped() ) {
+              data.two = getRowFrom( data.twoRowSet );
+            }
+
+            setOutputDone();
+            return false;
+          } else {
+            /*
+             * We are doing full outer join so print the 1st stream and get the next row from 1st stream
+             */
+            Object[] outputRowData = RowDataUtil.createResizedCopy( data.one, data.outputRowMeta.size() );
+            outputRowData = RowDataUtil.addRowData( outputRowData, data.oneMeta.size(), data.two_dummy );
+            putRow( data.outputRowMeta, outputRowData );
+            data.one = getRowFrom( data.oneRowSet );
+          }
+        } else if ( data.two == null && data.two_optional == true ) {
+          /**
+           * We have reached the end of stream 2 and there are records present in the first stream. Also, join is left
+           * or full outer. So, create a row with just the values in the first stream and push it forward
+           */
+          Object[] outputRowData = RowDataUtil.createResizedCopy( data.one, data.outputRowMeta.size() );
+          outputRowData = RowDataUtil.addRowData( outputRowData, data.oneMeta.size(), data.two_dummy );
+          putRow( data.outputRowMeta, outputRowData );
+          data.one = getRowFrom( data.oneRowSet );
+        } else if ( data.two != null ) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

It seems unusual to set the value of a variable only when it is already set. You might want to switch the null check of != on line 321 to == check before setting the field value on line 325 or verify if the check is needed at all.